### PR TITLE
Add Blip2 model in VQA pipeline

### PR DESCRIPTION
**Pull request by jpizarrom**
# What does this PR do?

Add Blip2ForConditionalGeneration model in VisualQuestionAnsweringPipeline.
Fixes part of #21110 and is based on #23348 #21227 .

## Who can review?

Hi @NielsRogge  what do you think of this??
Thanks!

## TODOs
- [x] add Blip2 model in VQA pipeline 
- [x] use require_torch_gpu in test
- [x] use can_generate in vqa pipeline for Blip2ForConditionalGeneration
- [x] use float16 in the test_large_model_pt_blip2
- [ ] check if it is necessary to cast the input in torch.float16 inside _forward

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @amyeroberts and @younesbelkada

**Comment by jpizarrom** [comment]
Hi @amyeroberts and @younesbelkada, this PR is ready for review. Thanks :)

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (younesbelkada) [inline tests/pipelines/test_pipelines_visual_question_answering.py:140]:
```suggestion
    @require_torch
    @require_torch_gpu
```

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (younesbelkada) [inline tests/pipelines/test_pipelines_visual_question_answering.py:142]:
```suggestion
        vqa_pipeline = pipeline("visual-question-answering", model="Salesforce/blip2-opt-2.7b", model_kwargs={"torch_dtype": torch.float16)
```
Make also sure to cast the input in `torch.float16` inside `_forward` if needed

**Review by younesbelkada — COMMENTED** [review]
Looking great to me, I left one comment to make sure the slow test will not blow up GPU memory in our daily CI runners! 
Let's also wait for amy and @Narsil 's review before merging

**Review by Narsil — APPROVED** [review]
LGTM.

I'm not a big fan of using the names of the configs directly to detect if generative or not, I feel like using the `ForXXX` should be a better hint.

We also used `model.can_generate()` as a hint in other pipelines.

Pinging @ylacombe who used that flag. (Just FYI no need to do anything).

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (jpizarrom) [reply tests/pipelines/test_pipelines_visual_question_answering.py]: @require_torch_gpu was added :)

**Comment by jpizarrom** [comment]
> LGTM.
> 
> I'm not a big fan of using the names of the configs directly to detect if generative or not, I feel like using the `ForXXX` should be a better hint.
> 
> We also used `model.can_generate()` as a hint in other pipelines.
> 
> Pinging @ylacombe who used that flag. (Just FYI no need to do anything).

@Narsil, Thanks a lot for your feedback :)

model.can_generate() return False for Blip2ForConditionalGeneration, that is the reason why I was following the proposal of this non merged PR https://github.com/huggingface/transformers/pull/23348/files#diff-620bada7977c3d0040ed961581379598e53a9ef02fdbb26c570cac738c279c0eR64

should/is it expected that can_generate method returns True for Blip2ForConditionalGeneration? if this is the case, we could use it.
If not, maybe the ForXXX flag could help, (what ForXXX flag do you mean?)

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (jpizarrom) [reply tests/pipelines/test_pipelines_visual_question_answering.py]: Hi @younesbelkada , thanks for your feedback :)
model_kwargs were updated in test_large_model_pt_blip2, as recommended by you, but i am not sure how to check if casting to `torch.float16 `inside `_forward` is needed, could you please give me some hints about what should I check? Thanks

**Comment by Narsil** [comment]
I'm not the best person to comment on how `can_generate` works and what should or shouldn't be done.

The main thing about pipeline:

- They should try to be model agnostic as much as possible so when newer models come in they work out of the box.

But the current code is acceptable.

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_visual_question_answering.py:199]:
```suggestion
        self.assertEqual(outputs, [[{"answer": "two"}]] * 2)
```

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (amyeroberts) [inline tests/pipelines/test_pipelines_visual_question_answering.py:180]: nit
```suggestion
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! 

Before approving let's get @gante's opinion on the best/canonical way to detect if the model should generate the answer or not within the pipeline

File: src/transformers/models/blip_2/modeling_blip_2.py
Comment (gante) [inline src/transformers/models/blip_2/modeling_blip_2.py:1797]: I'm going to ask to remove this change, and I'll add a more robust version myself -- in essence, classes with a custom `generate` should obviously return `True` here, and the general function does not indeed capture this case! 🤗

**Review by gante — COMMENTED** [review]
Regarding detection of generative models: `can_generate()` was built precisely to confirm whether the model can safely call `generate()` (in theory, all models can do it due to the inheritance structure, in practice only a few can use it). This includes pipelines uses 👍 

However, I don't think we should overload the function in the class -- see my comment below, going to open a PR with a more general solution :)

cc @amyeroberts @jpizarrom

**Comment by gante** [comment]
Generalizable solution here ☝️

File: src/transformers/models/blip_2/modeling_blip_2.py
Comment (gante) [reply src/transformers/models/blip_2/modeling_blip_2.py:1797]: `can_generate()` was updated, this can be safely removed (it now returns `True` for the model, without the overload)

File: tests/pipelines/test_pipelines_visual_question_answering.py
Comment (jpizarrom) [reply tests/pipelines/test_pipelines_visual_question_answering.py]: thanks, I have applied all your suggestions  :)

File: src/transformers/models/blip_2/modeling_blip_2.py
Comment (jpizarrom) [reply src/transformers/models/blip_2/modeling_blip_2.py]: Thanks a lot, custom can_generate was removed from [Blip2ForConditionalGeneration](https://github.com/huggingface/transformers/pull/25532/commits/d309aa4f913ee823f8f04a58177fc009d643e6af)

**Review by NielsRogge — APPROVED** [review]
Very clean and minimal PR, I like it!

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this functionality!

**Comment by NielsRogge** [comment]
Feel free to tweet about it @jpizarrom and we'll amplify :)

**Comment by RainyLayx** [comment]
I use the newest library of transformers,but it still reports "The model 'Blip2ForConditionalGeneration' is not supported for vqa. Supported models are ['ViltForQuestionAnswering'].",so how can I use blip2 in pipeline to deal with the vqa task?Are there any test codes of BLIP2 in pipeline?

**Comment by jpizarrom** [comment]
Hi @RainyLayx i was able to run this sample with _transformers==4.37.1_

```python
from transformers import pipeline
import requests
from PIL import Image

vqa_pipeline = pipeline("visual-question-answering", model="Salesforce/blip2-opt-2.7b")

image =  Image.open(requests.get("https://huggingface.co/datasets/Narsil/image_dummy/raw/main/parrots.png", stream=True).raw)
question = "Question: Is there a parrot? Answer:"

print(vqa_pipeline(image, question, top_k=1))
```

**Pull request merged**
